packages/chatbot-server-mongodb-public/src/processors/generateResponseWithTools.eval.ts

```diff
@@ -0,0 +1,288 @@
+
+type GenerateResponseExpectedMessage = {
+  role: SomeMessage["role"] | "assistant-tool";
+  toolCallName?: string;
+  toolCallArgs?: Record<string, string>;
+};
+
+type GenerateResponseExpected = {
+  messages: GenerateResponseExpectedMessage[];
+};
+
+// Load eval cases from a YAML file.
+const evalCases: EvalCase<
+  GenerateResponseInput,
+  GenerateResponseExpected,
+  IntermediateToolResponse
+>[] = getGenerateWithToolsEvalCasesFromYamlFile(
+  require.resolve("../../evalCases/generate_response_with_tools.yml")
+);
+
+/** Verify the correct tool calls were generated. */
+const ScoreCorrectToolsCalled: EvalScorer<
+  GenerateResponseInput,
+  GenerateResponseReturnValue,
+  GenerateResponseExpected
+> = ({ output, expected }) => {
+  if (output.messages.length !== expected.messages.length) {
+    return {
+      name: "CorrectToolCall",
+      score: 0,
+      metadata: {
+        message: "Output and expected messages length were different",
+      },
+    };
+  }
+
+  // Verify messages[i].toolCall is the same as expected.messages[i]
+  for (let i = 0; i < output.messages.length; i++) {
+    const expectedMessage = expected.messages[i];
+    const outputMessage = output.messages[i];
+
+    // Check if the message types match
+    if (
+      !(
+        expectedMessage.role === "assistant-tool" &&
+        outputMessage.role === "assistant"
+      ) &&
+      expectedMessage.role !== outputMessage.role
+    ) {
+      return {
+        name: "CorrectToolCall",
+        score: 0,
+        metadata: {
+          message: `Role mismatch at index ${i}: Expected ${expectedMessage.role}, got ${outputMessage.role}`,
+        },
+      };
+    }
+
+    // Check for expected assistant tool call by matching assistant role and toolCall presence
+    if (expectedMessage.role === "assistant-tool") {
+      const hasToolCall =
+        "toolCall" in outputMessage &&
+        outputMessage.toolCall &&
+        typeof outputMessage.toolCall === "object" &&
+        "function" in outputMessage.toolCall;
+
+      if (!hasToolCall) {
+        return {
+          name: "CorrectToolCall",
+          score: 0,
+          metadata: {
+            message: `Assistant did not return a tool call where one was expected`,
+          },
+        };
+      }
+      if (
+        expectedMessage.toolCallName !== outputMessage.toolCall?.function.name
+      ) {
+        return {
+          name: "CorrectToolCall",
+          score: 0,
+          metadata: {
+            message: `Assistant returned the wrong tool call`,
+          },
+        };
+      }
+    }
+  }
+  return {
+    name: "CorrectToolCall",
+    score: 1,
+  };
+};
+
+/** Verify the correct args were generated for the tool call(s) */
+const ScoreToolsUsedCorrectly: EvalScorer<
+  GenerateResponseInput,
+  GenerateResponseReturnValue,
+  GenerateResponseExpected
+> = ({ output, expected }) => {
+  let totalToolArgsCorrect = 0;
+  let totalToolArgs = 0;
+  for (
```

Comment:
the logic here seems fairly duplicative of the above scorer. 

i suggest creating one helper function that extracts all the necessary data for both the scorers, and then the logic in the scorer itself can be much simpler